modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/CacheGroupPageScanner.java

```diff
@@ -0,0 +1,567 @@
+            return cpFut.onDone(compFut.result());
+        }
+
+        /**
+         * Stop reencryption of the specified partition.
+         *
+         * @param partId Partition ID.
+         * @return {@code True} if reencryption was cancelled.
+         * @throws IgniteCheckedException If failed.
+         */
+        public boolean cancel(int partId) throws IgniteCheckedException {
+            IgniteInternalFuture<Void> fut = futMap.get(partId);
+
+            if (fut == null)
+                return false;
+
+            return fut.cancel();
+        }
+
+        /**
+         * @param t Throwable.
+         */
+        public void fail(Throwable t) {
+            cpFut.onDone(t);
+        }
+    }
+
+    /**
+     * Page store scanning task.
+     */
+    private class PageStoreScanTask extends GridFutureAdapter<Void> implements Runnable {
+        /** Partiion ID. */
+        private final int partId;
+
+        /** Cache group scan context. */
+        private final GroupScanContext scanCtx;
+
+        /**
+         * @param scanCtx Cache group scanning context.
+         * @param partId Partition ID.
+         */
+        public PageStoreScanTask(GroupScanContext scanCtx, int partId) {
+            this.scanCtx = scanCtx;
+            this.partId = partId;
+        }
+
+        /** {@inheritDoc} */
+        @Override public synchronized boolean cancel() throws IgniteCheckedException {
+            return onDone();
+        }
+
+        /** {@inheritDoc} */
+        @Override public void run() {
+            try {
+                int grpId = scanCtx.groupId();
+
+                CacheGroupContext grp = ctx.cache().cacheGroup(grpId);
+
+                if (grp == null) {
+                    onDone();
+
+                    return;
+                }
+
+                PageMemoryEx pageMem = (PageMemoryEx)grp.dataRegion().pageMemory();
+                long metaPageId = pageMem.partitionMetaPageId(grpId, partId);
+                long state = ctx.encryption().getEncryptionState(grpId, partId);
+                int batchSize = encrCfg.getReencryptionBatchSize();
+
+                int off = (int)(state >> Integer.SIZE);
+                int cnt = (int)state;
+
+                if (log.isDebugEnabled()) {
+                    log.debug("Partition reencryption is started [grpId=" + grpId +
+                        ", p=" + partId + ", remain=" + (cnt - off) + ", total=" + cnt + "]");
+                }
+
+                while (off < cnt) {
+                    int pagesCnt = Math.min(batchSize, cnt - off);
+
+                    if (limiter != null)
+                        limiter.acquire(pagesCnt);
+
+                    synchronized (this) {
+                        if (isDone())
+                            break;
+
+                        ctx.cache().context().database().checkpointReadLock();
+
+                        try {
+                            off += scanPages(pageMem, metaPageId + off, pagesCnt);
```

Comment:
Let's not rely on knowledge of pageId structure and pass just offset to scanPages and generate pageId each time from offset by page index.